pyro/infer/trace_elbo.py

```diff
@@ -23,7 +23,7 @@ def _compute_log_r(model_trace, guide_trace):
     for name, model_site in model_trace.nodes.items():
         if model_site["type"] == "sample":
             log_r_term = model_site["log_prob"]
-            if not model_site["is_observed"]:
+            if not model_site["is_observed"] and name in guide_trace.nodes:
```

Comment:
I may be forgetting something, but I thought `Trace_ELBO` requires the guide to include all model sites that are not observed. If that's the case, we wouldn't want to keep the old version where `Trace_ELBO` errors. Can you explain when a model site should not be in the guide?

Reply:
I agree with you and I cannot think of useful cases where this is the case. The point I make in the [issue](https://github.com/pyro-ppl/pyro/issues/3255) is that this triggers a warning when `has_rsample` is true and an error when it is false. I think they should both trigger a warning or both trigger an error. The suggested changes try to make the behavior consistent with the case `has_rsample = True`. But maybe it makes more sense to trigger an error everywhere?

Reply:
Ah, thanks for explaining, I think I better understand now.

I'd like to hear what other folks think (@martinjankowiak @eb8680 @fehiepsi).  One argument for erroring more often is that there is a lot code in Pyro that tacitly assumes all sites are either observed or guided.  I'm not sure what that code is, since we've only tacitly made that assumption, but it's worth thinking about: reparametrizers, `Predictive`, `AutoGuideMessenger`.

One argument for allowing "partial" guides is that it's just more general.  But if we decide to support "partial" guides throughout Pyro, I think we'll need to adopt importance sampling semantics, so we'd need to replace pyro's basic `Trace` data structure with a weighted trace, and replace sample sets with weighted sets of samples in a number of places e.g. `Predictive`.  This seems like a perfectly fine design choice for a PPL, but it is different from much of Pyro's existing usage, and I think we would need to make many small changes throughout the codebase including tutorials. 🤔